akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal
+
+import akka.actor
+import akka.actor.typed
+import akka.actor.typed.Behavior.{ SameBehavior, StoppedBehavior, UnhandledBehavior }
+import akka.actor.typed.{ ActorContext, ActorRef, Behavior, BehaviorInterceptor, ExtensibleBehavior, PreStartTarget, ReceiveTarget, Signal, SignalTarget, WrappingBehavior }
+import akka.annotation.InternalApi
+
+/**
+ * Provides the impl of any behavior that could nest another behavior
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object InterceptorImpl {
+
+  def apply[O, I](interceptor: BehaviorInterceptor[O, I], nestedBehavior: Behavior[I]): Behavior[O] = {
+    Behavior.DeferredBehavior[O] { ctx ⇒
+      val interceptorBehavior = new InterceptorImpl[O, I](interceptor, nestedBehavior)
+      interceptorBehavior.preStart(ctx)
+    }
+  }
+}
+
+/**
+ * Provides the impl of any behavior that could nest another behavior
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class InterceptorImpl[O, I](val interceptor: BehaviorInterceptor[O, I], val nestedBehavior: Behavior[I])
+  extends ExtensibleBehavior[O] with WrappingBehavior[O, I] {
+
+  // invoked pre-start to start/de-duplicate the initial behavior stack
+  def preStart(ctx: typed.ActorContext[O]): Behavior[O] = {
+    val started = interceptor.preStart(ctx.asInstanceOf[ActorContext[I]], new PreStartTarget[I] {
+      override def start(): Behavior[I] = {
+        Behavior.start[I](nestedBehavior, ctx.asInstanceOf[ActorContext[I]])
+      }
+    })
+    new InterceptorImpl(interceptor, started)
+  }
+
+  override def replaceNested(newNested: Behavior[I]): Behavior[O] =
+    new InterceptorImpl(interceptor, newNested)
+
+  override def receive(ctx: typed.ActorContext[O], msg: O): Behavior[O] = {
+    // FIXME can we avoid allocating one on each message?
```

Comment:
perhaps if adding `ActorContext` as parameter in the `ReceiveTarget` method

Reply:
or use the ctx from `preStart`, i.e. create the single `ReceiveTarget` instance in `preStart`

Reply:
If we pass the `ctx` interceptors that has different outer and inner types will have to typecast it themselves - not so nice.

The inner behavior can change, so we can't create a final one in preStart I think, but we could cache it, and compare whenever a signal or message is consumed.

Reply:
> The inner behavior can change

but the ActorContext remains the same

it's like if you use `setup`, that `ctx` can be used for the the entire life of the actor

Reply:
My suggestion about using preStart wasn't good because then the same interceptor instance can't be used in several different actors. I'll give the passing of context a try to see what the problem with that is.